ui: Search improvements

johncowen opened the pull request:
This PR is a collection of PRs to improve the search experience in the UI

- initial `phrase-editor` component (#5507)

pearkes commented on ui-v2/tests/acceptance/components/catalog-filter.feature:
```diff
@@ -60,7 +60,7 @@ Feature: components / catalog-filter
     -------------------------------------------------
     | Model   | Page     | Url                       |
-    | service | services | /dc-1/services            |
+    # | service | services | /dc-1/services            |
```
Is this intentional? It seems like this is testing the old UX? Perhaps this should just be removed? (same with the ignore statement).

pearkes replied:
Ah actually the other test is _just_ the component not an acceptance test, so seems like we're just missing update this for phrase editing?

johncowen replied:
Ah yeah, the comment should go, will update!

The `@ignore` will be there to pester me until I add a real test for it, it means I get a skipped message whilst running tests, although we are probably at a good point to get that in. Will shout when it's there 👍

pearkes commented on ui-v2/app/templates/dc/services/index.hbs:
```diff
@@ -10,7 +10,7 @@
     {{#block-slot 'toolbar'}}
 {{#if (gt items.length 0) }}
-        {{catalog-filter searchable=searchable search=filters.s status=filters.status onchange=(action 'filter')}}
+    {{#phrase-editor placeholder=(if (eq terms.length 0) 'service:name tag:name status:critical search-term' '') items=terms searchable=searchable}}{{/phrase-editor}}
```
Why is the conditional around the placeholder required? I would assume the browser managed that based on the content?

johncowen replied:
🤔 I think it's because the 'terms' (the little pill shapes) aren't actually in the input field, so only show the placeholder when `terms.length == 0`, otherwise you get a placeholder even if you have say 2 terms. I'll double check this though, but the more I read this the more I think that makes sense.

pearkes commented:
Something else I just noticed that we might want to think about...

Looks like the params change here:

https://demo.consul.io/ui/dc1/services?status=passing

So when you load it up against this branch you lose the filter:

https://ui-stage.consul.io/ui/dc1/services?status=passing

Is that concerning? I'm not sure if users will expect that or be bothered by that. Maybe we can do a shim  of some kind. Thoughts on impact there @opihana?

johncowen commented:
Oh! Good spot! Of course, because we removed the buttons, I should try and catch/shim that, lemme look.

johncowen commented:
Hey @pearkes ,

I added something to provide some backwards compat to those `status=passing` values (we now convert these to new style when the page loads up)

I also added some basic acceptance testing to a similar level to the previous UI component. Ideally we should add some more around here to cover the above case you caught also though.

Let me know what you think, potentially we could add some more testing further down the line.

Thanks,

John
